woocommerce/woocommerce
# Fix analytics report download link not being sent (action scheduler blocking)

### Changes proposed in this Pull Request:

Allow scheduler actions being run through trait `SchedulerTraits` to be blocked by other **delayed** jobs.

Please, check this video explaining the problem:

https://github.com/user-attachments/assets/cf8344ef-44a1-4909-b984-a70d16841942



### How to test the changes in this Pull Request:

1. Set up a vanilla WooCommerce environment (for example `pnpm wp-env start --xdebug`)
2. If necessary, disable wp-cron or otherwise make sure that action scheduler actions will be left in pending state so that they can be run manually
3. Make sure that the store has at least 26 products because analytics results will be paged after 25 products. We need them to be paged. 
![wc_analytics_26](https://github.com/user-attachments/assets/4aaa79f3-08f1-459a-8fef-3acb5e4e96d4)

4. Go to stock analytics page and click "Download": `/wp-admin/admin.php?page=wc-admin&path=%2Fanalytics%2Fstock`. Make sure you receive the message "Your Stock Report will be emailed to you."
![wc_stock_report_download](https://github.com/user-attachments/assets/ea0c465a-b6c5-472a-b7af-8a56c8ed020e)
![wc_stock_report_mailed](https://github.com/user-attachments/assets/ce94e4d6-9820-4590-b3f4-7040b498bc5a)

5. Make sure that two new pending scheduled actions were added: `/wp-admin/admin.php?page=wc-status&tab=action-scheduler&status=pending&orderby=hook&order=desc`
![wc_pending_jobs](https://github.com/user-attachments/assets/3d71fc59-be09-4fc6-8e38-6cf56e0e8bea)

6. Manually run the action `woocommerce_admin_email_report_download_link` by clicking "Run".
![wc_run_action](https://github.com/user-attachments/assets/bbd1e56a-9908-4a44-b5f7-983a10d1f892)

7. If everything's working correctly, the action `woocommerce_admin_email_report_download_link` should be re-scheduled because the action `woocommerce_admin_report_export` had not been run first. No email should have been sent.

8. Now run the `woocommerce_admin_report_export` action first and then the `woocommerce_admin_email_report_download_link` action.

9. Now, the download link action should **not** have been rescheduled and an email containing the download link **should** have been sent
![wc_report_download](https://github.com/user-attachments/assets/b96f7112-a36c-4fb0-9d6e-1fa9a2a798a9)



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
